### DISCUSSION TITLE
Tailwind gradients classes doesn't work when build Gridsome project

### DISCUSSION THREAD

**Discussion by alex-enchev** [tailwindlabs/tailwindcss#11639] [Help]
**What version of Tailwind CSS are you using?**
- "tailwindcss": "latest" (currently 3.3.3)

**What build tool (or framework if it abstracts the build tool) are you using?**
- Gridsome

**What version of Node.js are you using?**
- v14.21.3

**What browser are you using?**
- Chrome

**What operating system are you using?**
- macOS

**Reproduction URL**

https://tw.sekretar.bg/ - Super simple project only with gradient background classes.
https://github.com/Guliat/tw-issue - Public repository of the project

**Describe your issue**
If you inspect the div element from the project I provide you will see there are gradient classes, but the gradient is missing.
This is only when you build the project (gridsome build), When I work locally (gridsome develop) everything is ok, the classes works properly.
I research a lot, but without any success. I am not sure if is issue with PostCSS or with the Tailwind plugin for Gridsome.
I use Tailwind for all my projects (mostly VueJS) and I never have this issue, so I think is something  specific with Gridsome build process.

**Comment by adamwathan** [comment]
Hey! It looks like Gridsome is using a CSS minifier under the hood that incorrectly deletes important CSS, notable this:

```css
*, ::before, ::after {
  --tw-gradient-from-position:  ;
  --tw-gradient-via-position:  ;
  --tw-gradient-to-position:  ;
}
```

They are using `css-minimizer-webpack-plugin@^3.0.1` but the latest version is `5.0.1`:

https://github.com/gridsome/gridsome/blob/master/gridsome/package.json#L42

https://www.npmjs.com/package/css-minimizer-webpack-plugin

If you can find a way to use the latest version of cssnano under the hood it should resolve the issue for you 👍

**Reply by alex-enchev** [reply]
Hey! Thanks for the fast response!
I looked at this but in my "node_modules" gridsome's package.json is different.
"css-minimizer-webpack-plugin": "^3.0.1" is "optimize-css-assets-webpack-plugin": "^5.0.1"
... I read the documentations for this package and they say "⚠️ For webpack v5 or above please use [css-minimizer-webpack-plugin] instead." The "webpack" version in my node_modules is 4.29.3.
The strange thing is that I use the latest version of Gridsome and I have different version from this what you linked.

For now I think just to use Tailwind with the CLI watch method, because I have to spend much time to fix the Gridsome issue.
Maybe in next update it will be fixed by Gridsome's team.

Its seems that the issue is not related to Tailwind but to Gridsome so I want to say sorry to bother you and thanks for the replay :)
Have a nice day!

**Comment by ervinchai** [comment]
Hello, I've been facing a similar issue but I'm on Rails/Webpack:-
- Rails 7
- Webpacker 5
- Webpack 4

Here's how I resolved it:-

---

With Webpacker (and likely Gridsome), they are quite opinionated in the way they process CSS files. Webpacker is a Rails wrapper around Webpack, and provides defaults that works for most people (seems like we aren't most people 😢).

@Guliat Similarly, Webpacker uses the `optimize-css-assets-webpack-plugin`.

If we look closer at the `optimize-css-assets-webpack-plugin` package, we can see that it's actually using `cssnano`, and looking deeper, we can see that [this issue](https://github.com/cssnano/cssnano/issues/1165) describes exactly what @adamwathan mentioned about the custom variables.

---

With that said, since upgrading `cssnano` is not an option, we have to find a way to pass some options to `optimize-css-assets-webpack-plugin` to disable the `postcss-discard-empty` portion of it. This will cause the bundle size to be very slightly bigger, but you shouldn't be having empty CSS definitions in your original CSS file anyway.

Since I'm using `@rails/webpacker`, I needed to add this line of code into my `config/webpack/production.js`.
```
environment.plugins.get("OptimizeCSSAssets")["options"]["cssProcessorPluginOptions"] = {
    preset: ['default', { discardEmpty: false }]
}
```

If you're using Gridsome, you might have some differences in configuration structure, but the idea is to pass `preset: ['default', { discardEmpty: false }]` into `cssProcessorPluginOptions`. 

This effectively disables the `postcss-discard-empty` optimization on `cssnano`, which removes the important CSS variables that @adamwathan mentioned in their original post.

---

I've spent so much time wracking my head on this, and have never been happier to see a gradient in my life. Hope this helps anyone reading it!

<img width="521" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/88306776/6fd3eefb-1db0-42cf-a3a7-00c1edc09b11">

**Comment by alex-enchev** [comment]
@ervinchai I will try to implement this in Gridsome later
Currently I just run the "npx tailwindcss -i ./src/input.css -o ./dist/output.css --watch" and I use the generated .css file.
